internal/metrics/synthetics/agent.go

```diff
@@ -0,0 +1,176 @@
+package synthetics
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"encoding/json"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/blackbox_exporter/config"
+	"github.com/prometheus/blackbox_exporter/prober"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/superfly/flyctl/internal/logger"
+	"nhooyr.io/websocket"
+)
+
+func RunAgent(ctx context.Context) error {
+	logger := logger.FromContext(ctx)
+	logger.Info("running synthetics agent")
+	ws, err := NewMetricsWs()
+	if err != nil {
+		return err
+	}
+
+	if err = ws.Connect(ctx); err != nil {
+		logger.Errorf("error connecting to flynthetics: %v", err)
+		return err
+	}
+
+	go func() {
+		defer ws.wsConn.CloseNow()
+		c := make(chan os.Signal, 1)
+		signalChannel(c)
+
+		tick := time.NewTicker(5 * time.Second)
+		defer tick.Stop()
+
+		reconnectAt := time.Time{}
+
+		for {
+			select {
+			case <-tick.C:
+				if !reconnectAt.IsZero() && reconnectAt.Before(time.Now()) {
+					ws.lock.Lock()
+					ws.Connect(ctx)
+					ws.lock.Unlock()
+
+					reconnectAt = time.Time{}
+				}
+
+			case <-ctx.Done():
+				return
+
+			case <-c:
+				if reconnectAt.IsZero() {
+					reconnectAt = time.Now().Add(5 * time.Second)
+				}
+
+			case <-ws.reset:
+				if reconnectAt.IsZero() {
+					reconnectAt = time.Now().Add(5 * time.Second)
+				}
+			}
+		}
+	}()
+
+	go listen(ctx, ws)
+
+	return nil
+}
+
+func listen(ctx context.Context, ws *SyntheticsWs) error {
+	logger := logger.FromContext(ctx)
+	logger.Debug("start listening for probes")
+	for ctx.Err() == nil {
+		ws.lock.RLock()
+		c := ws.wsConn
+		ws.lock.RUnlock()
+
+		_, message, err := c.Read(ctx)
+		if err != nil {
+			logger.Error("read error: ", err)
+			ws.resetConn(c, err)
+			continue
+		}
+
+		logger.Debug("received from server", string(message))
+
+		err = processProbe(ctx, message, ws)
+		if err != nil {
+			logger.Error("failed processing probe", err)
+		}
+
+	}
+	logger.Debug("stop listening for probes")
+
+	return ctx.Err()
+}
+
+type ProbeMessage struct {
+	Module string `json:"module"`
+	Target string `json:"target"`
+}
+
+func processProbe(ctx context.Context, jsonMessage []byte, ws *SyntheticsWs) error {
+	logger := logger.FromContext(ctx)
+	logger.Debug("proccessing probes")
+
+	probeMessage := ProbeMessage{}
+	err := json.Unmarshal(jsonMessage, &probeMessage)
+	if err != nil {
+		logger.Error("JSON parse error:", err)
+		return err
+	}
+
+	probeSuccessGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_success",
+		Help: "Displays whether or not the probe was a success",
+	})
+	probeDurationGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_duration_seconds",
+		Help: "Returns how long the probe took to complete in seconds",
+	})
+
+	module := config.Module{}
+	module.HTTP.IPProtocol = "ipv4"
```

Comment:
Thanks for finding this, I overlooked to remove the hardcoded protocol (I only have ipv4 😄 )
 
The blackbox exporter behaves like this [by design](https://github.com/prometheus/blackbox_exporter/issues/594), it will default to ipv6 for DNS resolution of the target, but if there is no ipv6 route the probe is considered failed instead of using ipv4 as fallback.

I adjusted this to receive the preferred/available protocol from the server side, as we can get it from the Websocket connection (server [PR](https://github.com/superfly/flynthetics/pull/1) )

Reply:
Worth noting that all the configuration aspect for the probe is still pending. This configuration is too restrictive, probably useless for customers. Future iterations will need to take the module configuration from the probe request. If we finally open this functionality to customers, we'd need to be flexible about the http method, success response codes, headers, etc.